File: leetcode/459.重复的子字符串/459.cpp
class Solution {
public:
    void getNext(vector<int>& next, string& s) {
        int prefix_len = 0, i = 1;
        while (i < s.size()) {
            if (s[prefix_len] == s[i]) {
                next[i++] = ++prefix_len;
            }
            else {
                if (prefix_len == 0) {
                    next[i++] = 0;
                }
                else {
                    prefix_len = next[prefix_len - 1];
                }
            }
        }
    }
    bool repeatedSubstringPattern(string s) {
        vector<int> next(s.size());
        next[0] = 0;
        getNext(next, s);
        int len = s.size();
        if (next[len - 1] != 0 && len % (len - next[len - 1]) == 0) {
            return true;
        }
        return false;
    }
};

File: leetcode/459.重复的子字符串/459.py
class Solution:
    def getNext(self, next, s):
        prefix_len, i = 0, 1
        while i < len(s):
            if s[prefix_len] == s[i]:
                prefix_len += 1
                next[i] = prefix_len
                i += 1
            else:
                if prefix_len == 0:
                    next[i] = 0
                    i += 1
                else:
                    prefix_len = next[prefix_len - 1]

    def repeatedSubstringPattern(self, s: str) -> bool:
        next = [0] * len(s)
        self.getNext(next, s)
        l = len(s)
        if(next[l - 1] != 0 and l % (l - next[l - 1]) == 0):
            return True
        return False